### ISSUE TITLE
Infinite redirects after login

### ISSUE THREAD

**Issue by venkatmarepalli** [zmartzone/lua-resty-openidc#130]
When I updated our version from 1.4.0 to 1.5.2, I see that after login the page never redirects back to the home page. It continuously redirects back and forth to Login page. We use Microsoft Azure B2C for federation.

Previously when user types https://homepage.com openidc takes them to Microsoft login page and once logged in redirects back to the https://homepage.com. Looks like something is broken for us. 

Has anyone faced similar issue?

**Comment by bodewig** [comment]
Probably token verification fails, is there any message logged?

**Comment by venkatmarepalli** [comment]
I don't see any error msgs.

**Comment by zandbelt** [comment]
browser traces and a server debug log would help to analyze this

**Comment by venkatmarepalli** [comment]
Here is the browser trace log. 
[Browser Trace.Log](https://github.com/zmartzone/lua-resty-openidc/files/1598876/Browser.Trace.Log)

**Comment by zandbelt** [comment]
I see two issues with the authorization request:
1. the response type used is "id_token" whereas lua-resty-openidc supports only "code"
2. the response mode is "form_post" which is not (yet) supported by lua-resty-openidc

I don't see how this would have worked with 1.4.0.

Apparently you're setting response mode/type in your config. Can you paste your NGINX conf?

**Comment by zandbelt** [comment]
I see there's no access token in the authentication response which probably has to do with the observed behavior.

Could it be that you've set `renew_access_token_on_expiry` in your NGINX config which then forces reauthentication when there's no access token returned?.

In any case your OP returns an invalid authentication response since the OAuth 2.0 spec (which is what OpenID Connect is built on) requires the presence of an access token in the authentication response. Is there any way you can make Azure B2C return a spec compliant response?

**Comment by venkatmarepalli** [comment]
Azure AD B2C has a feature where we can customize the login page, the id_token urls that we are seeing in the logs are intermediate urls. The initial url has responce_type=code.  The flow goes like this 
1. User navigates to App Page https://myapp.com
2. openidc forms OAuth url with client_id etc. https://account.microsoft.com/lua_params.....
3. The login page above has some javascript which behaves differently for Internal users and external users based on the entered emailid's domain.
4. Redirects to different page for internal user and completes authentication and comes back to the original url which in turn is handled by openidc.

I am attaching my openidc.lua 1.4.0 (Patched version) and the server debug log generated using this version. I don't have any issues with 1.4.0 version.

[openidc-1.4.0.txt](https://github.com/zmartzone/lua-resty-openidc/files/1601307/openidc-1.4.0.txt) - Change extension to .lua
[nginx_error_psgi.log](https://github.com/zmartzone/lua-resty-openidc/files/1601354/nginx_error_psgi.log)
Nginx Config file -- [nginx_psgi.txt](https://github.com/zmartzone/lua-resty-openidc/files/1601379/nginx_psgi.txt)

**Comment by zandbelt** [comment]
Can you use the following patch to provide  more info on why re-authentication starts?
(I'll add this to the master branch later)
```
diff --git a/lib/resty/openidc.lua b/lib/resty/openidc.lua
index e881577..7d03d86 100644
--- a/lib/resty/openidc.lua
+++ b/lib/resty/openidc.lua
@@ -1006,6 +1006,16 @@
     end
   end
 
+  ngx.log(ngx.DEBUG,
+    "session.present=", session.present,
+    ", session.data.id_token=", session.data.id_token ~= nil,
+    ", session.data.authenticated=", session.data.authenticated,
+    ", opts.force_reauthorize=", opts.force_reauthorize,
+    ", opts.renew_access_token_on_expiry=", opts.renew_access_token_on_expiry,
+    ", try_to_renew=", try_to_renew,
+    ", token_expired=", token_expired
+  )
+  
   -- if we are not authenticated then redirect to the OP for authentication
   -- the presence of the id_token is check for backwards compatibility
   if not session.present
```

**Comment by venkatmarepalli** [comment]
`session.present=true, session.data.id_token=true, session.data.authenticated=true, opts.force_reauthorize=nil, opts.renew_access_token_on_expiry=nil, try_to_renew=true, token_expired=true, client: 172.17.0.1, server: localhost., request: "GET / HTTP/1.1", host: "xyz-qa-ext-blue.xyz.com", referrer: "https://login.microsoftonline.com/directory/oauth2/authorize?client_id=111&redirect_uri=https%3a%2f%2faccount.xyz.com%2aaaa-8614-adea42784599%2foauth2%2fauthresp&response_type=id_token&response_mode=form_post&nonce=2kwfhIrgpK5yIE9ob8fNCw%3d%3d&state=StateProperties%3deyJTSUQiOiJ4LW1zLWNwaW0tcmM6ZTAyMTczZGEtNjVhNC00MjE4LWIzNGMtNjJmMTk3YzQzNDJjIiwiVElEIjoiZjIzMjFlMDUtMmUzNS00OTBhLWIwZjEtMmI3YTRhMDIwMzk2In0"
`

**Comment by zandbelt** [comment]
setting `opts.renew_access_token_on_expiry` explicitly to `false` would be a workaround for this non-compliant OP behavior

**Comment by venkatmarepalli** [comment]
That worked !!

**Comment by zandbelt** [comment]
in your case there is no access token so nothing happens

when there is an access token and it is expired, lua-resty-openidc will try to renew on the backchannel with the `refresh` grant type it if `opts.renew_access_token_on_expiry` is not set to `false`; in case renewing it fails, the user will be asked to login again

**Comment by ali-panahi** [comment]
Hi , Can I ask Mr. @zandbelt about something?
I have a bit of a problem.
I've configured keycloak , nginx on openresty and It looks properly.
But after I try to access the http://[URL ADDRESS]:8080 , which I defined it in nginx.conf with lua , the login page of keycloak appears , but after successful login IE shows the error :
response indicates failure, status=403, body={"error":"invalid_request","error_description":"HTTPS required"}
my keycloak is http and I don't want to user https at all.
